qiskit_optimization/algorithms/cplex_optimizer.py

```diff
@@ -116,33 +139,25 @@ def solve(self, problem: QuadraticProgram) -> OptimizationResult:
             QiskitOptimizationError: If the problem is incompatible with the optimizer.
         """
 
-        # convert to CPLEX problem
-        cplex = problem.to_docplex().get_cplex()
-
-        # set display setting
-        if not self.disp:
-            cplex.set_log_stream(None)
-            cplex.set_error_stream(None)
-            cplex.set_warning_stream(None)
-            cplex.set_results_stream(None)
-
-        # solve problem
-        try:
-            cplex.solve()
-        except CplexSolverError as ex:
-            raise QiskitOptimizationError(str(ex)) from ex
-
-        # process results
-        sol = cplex.solution
-
-        # create results
-        result = OptimizationResult(
-            x=sol.get_values(),
-            fval=sol.get_objective_value(),
-            variables=problem.variables,
-            status=self._get_feasibility_status(problem, sol.get_values()),
-            raw_results=sol,
-        )
-
-        # return solution
+        mod = problem.to_docplex()
+        sol = mod.solve(log_output=self._disp, cplex_parameters=self._cplex_parameters)
+        if sol is None:
+            warn("CPLEX cannot solve the model")
+            x = [0.0] * mod.number_of_variables
+            result = OptimizationResult(
+                x=x,
+                fval=problem.objective.evaluate(x),
+                variables=problem.variables,
+                status=OptimizationResultStatus.FAILURE,
+                raw_results=None,
+            )
+        else:
+            x = sol.get_values(mod.iter_variables())
+            result = OptimizationResult(
+                x=x,
+                fval=sol.get_objective_value(),
+                variables=problem.variables,
+                status=self._get_feasibility_status(problem, x),
+                raw_results=sol,
+            )
```

Comment:
May be the code would look better if you pulled the `result` creation out of the `if` statement? This is just an idea, not a suggestion.

Reply:
It might be difficult to simplify it because the values of parameters are different in `if` statement. How about returning directly `OptimizationResult`, i.e., `return OptimizationResult{...}` in both `if` and `else` statements.

Reply:
OK, fine with me!